### DISCUSSION TITLE
'Fixing'/updating Homebrew after changing shells?

### DISCUSSION THREAD

**Discussion by kenny-evitt** [Homebrew/discussions#2700] [Everyday usage]
I installed Homebrew with the zsh shell and then installed some packages.

Then I switched zsh to Bash.

I re-ran the commands listed in the "Next steps" output of the installer but changed the 'shell profile' file name:

```
$ echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> /Users/kenny/.bash_profile
$ eval "$(/opt/homebrew/bin/brew shellenv)"
```

I noticed that in the "Caveats" output for one of the packages I installed it explicitly mentions "zsh completions" but no other shells.

Is there anything I *can* do – or *should* do – besides the above two commands? Re-install (all) packages? Create a `Brewfile` using `brew bundle dump` and then un-install and re-install Homebrew, and finally run `brew bundle` again?

**Comment by gromgit** [comment]
> Is there anything I _can_ do – or _should_ do – besides the above two commands?

Nope, that's pretty much it. Switching shells just requires setting up the new *shell*'s *env*ironment. As the subcommand name implies, the `brew shellenv` line does just that.

> I noticed that in the "Caveats" output for one of the packages I installed it explicitly mentions "zsh completions" but no other shells.

https://docs.brew.sh/Shell-Completion would be a useful read, if you need shell completions.

**Reply by kenny-evitt** [reply]
@gromgit Thanks!

So shell completions specifically should be possible to setup as your link describes – that's nice! And I just checked – one of the packages I installed (with completions) *does* have a file under the Bash completions directory. I wasn't sure whether Homebrew installed them given that the then-current shell was zsh (and Homebrew only mentioned zsh when I installed the package).

Given that the 'caveats' for packages seems specific to, or maybe 'filtered' by, the shell used when a package is installed, I was thinking that there might be others beyond shell completions and that there might be some way to see them for a 'new' shell – but without re-installing every package (or also uninstalling/re-installing Homebrew too).

**Reply by gromgit** [reply]
> I wasn't sure whether Homebrew installed them given that the then-current shell was zsh (and Homebrew only mentioned zsh when I installed the package).

Homebrew always installs _everything_ in the package. Regardless of what shell you're running, the available completions for _all_ supported shells will always be installed.

> Given that the 'caveats' for packages seems specific to, or maybe 'filtered' by, the shell used when a package is installed

The currently-running shell has no effect on caveats printed.

**Reply by kenny-evitt** [reply]
> The currently-running shell has no effect on caveats printed.

So the caveats for that package only mentioned the shell I was using and not another shell for which it did install the same thing (completions)? Would that be a bug in the package/formula then?

**Reply by gromgit** [reply]
No. If you're running bash, and no bash-specific caveats appear, it literally means that none were written.

You never did mention which formula you were installing, so I'll just pick one at random:
```
bash$ brew cat git-extras
class GitExtras < Formula
[...]
  def caveats
    <<~EOS
      To load Zsh completions, add the following to your .zshrc:
        source #{opt_pkgshare}/git-extras-completion.zsh
    EOS
  end
[...]

bash$ brew info git-extras
git-extras: stable 6.3.0 (bottled), HEAD
Small git utilities
https://github.com/tj/git-extras
[...]
==> Caveats
To load Zsh completions, add the following to your .zshrc:
  source /usr/local/opt/git-extras/share/git-extras/git-extras-completion.zsh
[...]

bash$ find /usr/local/opt/git-extras/ -name \*completion*
/usr/local/opt/git-extras//etc/bash_completion.d
/usr/local/opt/git-extras//share/git-extras/git-extras-completion.zsh
```
Notice how `brew info` shows me Zsh-specific caveats, even though I'm running bash _and_ there are bash completions available. Caveats are driven entirely by what's coded in the formula.

**Reply by kenny-evitt** [reply]
One particular formula I installed that seemed to demonstrate the alleged behavior is `pass`.

I installed it when I was still using zsh; here's the history of that:

```
% brew install pass
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> New Formulae
lighthouse                                                                                                                                                                                                          teku
==> Updated Formulae
Updated 93 formulae.

==> Downloading https://ghcr.io/v2/homebrew/core/gnu-getopt/manifests/2.37.2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/gnu-getopt/blobs/sha256:5fe2a8f7dd7c5b369f4e8ffce9bbc4d3f3d1a83bdf7ecf592757e101b34c4023
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:5fe2a8f7dd7c5b369f4e8ffce9bbc4d3f3d1a83bdf7ecf592757e101b34c4023?se=2021-12-29T01%3A00%3A00Z&sig=jc5%2BFPnC6UM0WdQunhLuMJ84t86vi4ucetk%2FRO4mTyI%3D&sp=r&spr=https&sr=b&sv=2019-12-12
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libpng/manifests/1.6.37
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/libpng/blobs/sha256:40b9dd222c45fb7e2ae3d5c702a4529aedf8c9848a5b6420cb951e72d3ad3919
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:40b9dd222c45fb7e2ae3d5c702a4529aedf8c9848a5b6420cb951e72d3ad3919?se=2021-12-29T01%3A00%3A00Z&sig=x8CWLFSfhMOrdvYvPhMOhkllPmjFNTpUYYa3qhnNzzA%3D&sp=r&spr=https&sr=b&sv=2019-12-12
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/qrencode/manifests/4.1.1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/qrencode/blobs/sha256:6fa3a670e9708cf84470f82fd966e5610d0ad9d8c96c1f5987645b4db3fa65cb
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:6fa3a670e9708cf84470f82fd966e5610d0ad9d8c96c1f5987645b4db3fa65cb?se=2021-12-29T01%3A00%3A00Z&sig=yFrPz1JVTgcXPU2Aou5NT6rU1kwMYHpAMh21ET6mbCc%3D&sp=r&spr=https&sr=b&sv=2019-12-12
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/tree/manifests/2.0.0_1
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/tree/blobs/sha256:9eb668c56cb2a8dde9d6090addaf8373c62321699d4a56c6eac3dab159132b5c
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:9eb668c56cb2a8dde9d6090addaf8373c62321699d4a56c6eac3dab159132b5c?se=2021-12-29T01%3A00%3A00Z&sig=UGzYyhROPKrvraPdD3h7zSikVzP9WGqISRB4tVmg5MI%3D&sp=r&spr=https&sr=b&sv=2019-12-12
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pass/manifests/1.7.4
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/pass/blobs/sha256:7ee1c46e172270c3161ef1d843c37f0bc21fb874af0b55dcd16ad32e795d17e5
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:7ee1c46e172270c3161ef1d843c37f0bc21fb874af0b55dcd16ad32e795d17e5?se=2021-12-29T01%3A00%3A00Z&sig=Sa2Zhw3baFNhLF6VEDJAm8Yo0itzHQdlVUVhkzLt6lQ%3D&sp=r&spr=https&sr=b&sv=2019-12-12
######################################################################## 100.0%
==> Installing dependencies for pass: gnu-getopt, libpng, qrencode and tree
==> Installing pass dependency: gnu-getopt
==> Pouring gnu-getopt--2.37.2.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/gnu-getopt/2.37.2: 10 files, 207.4KB
==> Installing pass dependency: libpng
==> Pouring libpng--1.6.37.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/libpng/1.6.37: 27 files, 1.3MB
==> Installing pass dependency: qrencode
==> Pouring qrencode--4.1.1.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/qrencode/4.1.1: 13 files, 308.9KB
==> Installing pass dependency: tree
==> Pouring tree--2.0.0_1.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/tree/2.0.0_1: 8 files, 166.7KB
==> Installing pass
==> Pouring pass--1.7.4.arm64_monterey.bottle.tar.gz
==> Caveats
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions

Emacs Lisp files have been installed to:
  /opt/homebrew/share/emacs/site-lisp/pass
==> Summary
🍺  /opt/homebrew/Cellar/pass/1.7.4: 34 files, 156.3KB
==> Running `brew cleanup pass`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
==> Caveats
==> pass
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions

Emacs Lisp files have been installed to:
  /opt/homebrew/share/emacs/site-lisp/pass
```

There's no mention of "bash" at all and yet the *Bash* completion does seem to have been installed:

```
$ ls /opt/homebrew/etc/bash_completion.d/
brew	pass
```

Based on what you wrote, that would be a bug in the `pass` formula, right?

**Reply by gromgit** [reply]
No, sorry, the bug's in my own firmware. There _is_ current-shell autodetect code that drives autogeneration of completion-related caveats here: https://github.com/Homebrew/brew/blob/a0ae4a7a02d9b6d886fb99a23ddbe07af7979fb6/Library/Homebrew/caveats.rb#L30-L41

It didn't trigger in the VM I tested because I'd checked out an older Homebrew revision and forgotten to switch back to master. 😁 

As for what you're seeing, that's almost certainly because your `SHELL` environment variable is still pointing to Zsh. To confirm:
```
echo $SHELL
SHELL= brew info pass
```
The second command, by temporarily nuking `SHELL`, forces Homebrew to actually check the parent (shell) process via `ps`, which is significantly more expensive than looking at an environment variable. From bash, the second command should tell you about the bash completions, but not the Zsh ones.

If you "switched" to Bash by simply running `bash`, note that `SHELL` points to your _login_ shell by convention, so you might want to `sudo chsh -s /bin/bash`, then re-login to fix that.

**Reply by kenny-evitt** [reply]
@gromgit I really appreciate all of your help – thanks! (And happy new year!)

My `SHELL` variable seems fine:

```
$ echo "$SHELL"
/bin/bash
```

I think, given what you just discovered about Homebrew auto-detecting the current shell for generating completion-related caveats entirely explains what I observed.

I didn't think to try `brew info pass`. That *does* (now) show the Bash completion:

```
$ brew info pass
pass: stable 1.7.4 (bottled), HEAD
...
==> Caveats
Bash completion has been installed to:
  /opt/homebrew/etc/bash_completion.d

Emacs Lisp files have been installed to:
  /opt/homebrew/share/emacs/site-lisp/pass
...
```

Thanks again!
